# HW 4
__CS 216, Everything Data, Spring 2020__

__DUE: Monday Feb. 17 by 4:40 pm (class time)__

__NAMES: Pierce Forte (phf7), Joe Cusano (jgc28)__

In this assignment, you will practice some statistical and probabilistic arguments that are useful for exploratory data analysis. You will include all of your answers for this assignment within this notebook. You will then convert your notebook to a .pdf and a .py file to submit to gradescope (submission instructions are included at the bottom).

Please take note of the [course collaboration policy](https://sites.duke.edu/compsci216s2020/policies/). You may work alone or with a single partner. If you work with a partner, you may not split up the assignment; you should work together in-person or complete parts independently and come together to discuss your solutions. In either case, you are individually responsible for your work, and should understand everything in your submission.

# Part 1: Team Probabilities
Suppose that we take 150 students and randomly assign them to groups of 5 students each by the following procedure. First, we generate a uniform random permutation of the students (that is, a random ordering where for a given student there is an equal probability of appearing at any position in the ordering from 1, 2, 3, ..., 150). Then we simply assign the first five students in the ordering to the first group, the second five to the second group, and so on. Compute the following probabilities.

Note that you can arrive at your answer and show your work either analytically (show your math) or computationally (show code for your simulation). With the latter approach, you simulate the process many times, and see what fraction of these times the event of interest occcurs. In either case, we will count your answer as correct if it is correct rounded to the second decimal place (that is, if the real answer is 0.0312 and you get 0.0297, we will count that as correct).

### Problem A
Suppose a student has a single best friend in the class. What is the probability that a given student gets placed in a group with their best friend?

In [1]:
import random
import numpy as np

def randomlyAssignGroups():
    students = list(range(1, 151))
    random.shuffle(students)

    groups = []
    for set_of_5 in range(0,150//5):
        group = []
        for k in range(0,5):
            group.append(students[(set_of_5*5 + k)])
        groups.append(group)
    return groups

def inSameRandomGroup(person_a, person_b, groups):
    for group in groups:
        if person_a in group and person_b in group:
            return True
    return False
    
person = 1
friend = 2

sameGroup = 0
total = 0
for i in range(0, 100001):
    total+=1
    groups = randomlyAssignGroups()
    if inSameRandomGroup(person, friend, groups):
        sameGroup+=1
        
print(100*(sameGroup/total))


2.8269717302826973


We used code to determine that the probability is around 3%.  

### Problem B
Suppose a student has a single worst enemy in the class. What is the probability that a given student is *not* placed in a group with their worst enemy?

In [2]:
person = 1
enemy = 2

diffGroup = 0
total = 0
for i in range(0, 10001):
    total+=1
    groups = randomlyAssignGroups()
    if not inSameRandomGroup(person, friend, groups):
        diffGroup+=1
        
print(100*(diffGroup/total))

97.28027197280272


We used code to determine that the probability is around 97%.  This makes sense, as the probability from the previous question and this question should add up to 100%.  3% + 97% = 100%.

### Problem C
Suppose that the best friend relationship is symmetric: if Bob is Alice's best friend, then Alice is Bob's best friend; there are exactly 75 such best friend pairs. What is the expected number of best friend pairs that get placed into the same group accross all of the groups?

In [3]:
students = list(range(1, 151))
pairs = []
for set_of_2 in range(0,150//2):
        pair = []
        for k in range(0,2):
            pair.append(students[(set_of_2*2 + k)])
        pairs.append(pair)

def numPairsInSameGroup(pairs):
    numPairsInSameGroup = 0
    groups = randomlyAssignGroups()
    for pair in pairs:
        if inSameRandomGroup(pair[0], pair[1], groups):
            numPairsInSameGroup+=1
    return numPairsInSameGroup
    

listOfNumPairsInSameGroup = []

for i in range(0, 1001):
    listOfNumPairsInSameGroup.append(numPairsInSameGroup(pairs))

print(np.mean(listOfNumPairsInSameGroup)) # Should this be mode?

2.0669330669330668


We used code to determine that the expected number of best friend pairs that get placed into the same group across all groups should be around 2.

### Problem D
What is the probability that every best friend pair gets placed into the same group? (*hint* Don't overthink it).

In [4]:
numEveryPairInSameGroup = 0
total = 0

for i in range(0, 10001):
    total +=1
    if numPairsInSameGroup(pairs) == 75:
        numEveryPairInSameGroup +=1

print(100*(numEveryPairInSameGroup/total))

0.0


We used code to determine that the probability that every best friend pair gets placed into the same group is extremely small (around 0%).

### Problem E 
What is the probability that no best friend pairs get placed into the same group?

In [5]:
numNoPairsInSameGroup = 0
total = 0

for i in range(0, 100001):
    total +=1
    if numPairsInSameGroup(pairs) == 0:
        numNoPairsInSameGroup +=1

print(100*(numNoPairsInSameGroup/total))

12.974870251297489


The probability that no best friend pairs are placed into the same group is around 13%.

### Problem F
Now suppose that we couldn't get a big enough room, and there are 5 sections of the class that meet at different times, each with 30 students. Suppose we randomly assign groups as before, but only within sections so that there are 6 random groups with students from the first section, 6 random groups with students from the second section, and so on.

Reasonably enough, students may prefer to sign up for sections with their best friends. Suppose that 50 out of the 75 best friend pairs signed up for the same sections, and 25 of the best friend pairs had to take different sections. Given this information, what is the expected number of best friend pairs that get placed into the same group accross all of the groups?

In [6]:
def assignSections():
    sections = [[],[],[],[],[]]

    for i in range (0, 50):
        section = (i*2) // 20
        for student in pairs[i]:
            sections[section].append(student)
    
    for i in range (50, 75):
        for student in pairs[i]:
            section = student%5
            sections[section].append(student)
            
    return sections

def randomlyAssignGroupsInSection():
    section_groups = [[],[],[],[],[]]
    sections = assignSections()
    for section in sections:
        random.shuffle(section)
    groups = []
    
    for i in range(0,len(sections)):
        section = sections[i]

        for set_of_5 in range(0,30//5):
            group = []
            for k in range(0,5):
                group.append(section[(set_of_5*5 + k)])
            section_groups[i].append(group)
        
    return section_groups

def numPairsInSameSectionGroup(pairs):
    numPairsInSameGroup = 0
    section_groups = randomlyAssignGroupsInSection()
    for pair in pairs:
        for section in section_groups:
            if inSameRandomGroup(pair[0], pair[1], section):
                numPairsInSameGroup+=1
    
    return numPairsInSameGroup


listOfNumPairsInSameGroup = []
for i in range(0, 1001):
    listOfNumPairsInSameGroup.append(numPairsInSameSectionGroup(pairs))

print(np.mean(listOfNumPairsInSameGroup))

6.8551448551448555


The expected number of best friends that get placed into the same group is around 7.

# Part 2: Birthmonth Probabilities
To simplify this problem, assume that a year has twelve equal duration months, rather than the unequal months in the standard calendar. As before, show your work, either mathematical reasoning or simulation code.

### Problem A
Assuming that people are equally likely to be born in any month, how many people must be selected uniformly at random before the probability that at least two of the selected people have the same birth month is at least 1/2?

In [7]:
import random
same = 0
runs = 1000000
months = ["January", "February", "March", "April", "May", "June", "July", "August", "September", "October", "November", "December"]
for i in range(runs):
    selections = 2
    selected = []
    for i in range(selections):
        selected.append(random.choice(months))
    mySet = set(selected)
    if len(mySet) < len(selected):
        same += 1
r = same / runs
print(r)

0.083364


When 5 people are selected, the odds that at least two of the people have the same birthmonth is around .619.  When 4 people are selected, the odds that two people have the same birthmonth is around .428.  Thus, 5 people must be selected at random for the probability that at least two of the selected people to have the same birth month be at least 1/2. I altered the number of selections using the "selections" variable.

### Problem B
Now assume that for each month among May, June, July, and August, the probability of being born in that month is 1/10, whereas for every other month the probability of being born in that month is 3/40. What is the probability that two people selected uniformly at random have the same birth month? Is this more or less than in problem A?

In [8]:
import random
same = 0
runs = 1000000
months = ["January", "February", "March", "April", "May", "June", "July", "August", "September", "October", "November", "December"]
specialMonths = ["May", "June", "July", "August"]
editedMonths = []
for month in months:
    if month in specialMonths:
        for i in range(4):
            editedMonths.append(month)
    else:
        for i in range(3):
            editedMonths.append(month)
for i in range(runs):
    selections = 2
    selected = []
    for i in range(selections):
        selected.append(random.choice(editedMonths))
    mySet = set(selected)
    if len(mySet) < len(selected):
        same += 1
r = same / runs
print(r)

0.085061


When two people are selected at random, the probability that they have the same birth month is around .083 when the probability of being born in any given month is uniform.  When two people are selected at random the probability of being born in the same month is around 0.085 when the probability of being born in any given month is not uniform.  Thus, there is a slightly greater probability in the second scenario.  However, this greater probability could potentially be atrriubuted to the variability of the simulation.

# Part 3: Exploring Movie Lens Data
The movie lens dataset contains about 100,000 ratings from 1000 users on 1700 movies, made available by grouplen.org. This information is in three files: `u.user`, `u.data`, and `u.item`. Below, we import the relevant information into three dataframes (about the users, the ratings, and the movies). You are welcome to work directly with these dataframes, or to transform the data however you like for solving the problems in this part. Most of the data is self explanatory, but be sure to note that a given movie can be categorized as multiple genres (see the movies dataframe / `u.item` file below). To denote this, each row of the table corresponds to a given movie, and there is a column for every genre; there is a 1 in the column if the movie is of that genre, and a 0 otherwise.

In [9]:
import pandas as pd

df_users = pd.read_table('u.user', sep='|', names = ['user_id', 'age', 'sex', 'occupation', 'zip'])
df_users.head()

user_id  age sex  occupation    zip
0        1   24   M  technician  85711
1        2   53   F       other  94043
2        3   23   M      writer  32067
3        4   24   M  technician  43537
4        5   33   F       other  15213

In [10]:
df_ratings = pd.read_table('u.data', sep=',', names = ['user_id', 'movie_id', 'rating', 'timestamp'])
df_ratings.head()

user_id  movie_id  rating  timestamp
0      196       242       3  881250949
1      186       302       3  891717742
2       22       377       1  878887116
3      244        51       2  880606923
4      166       346       1  886397596

In [11]:
df_movies = pd.read_table('u.item', sep='|', encoding='latin',\
                          names = ['movie_id', 'movie_title', 'release_date', 'video_release_date',\
                                    'imdb_url', 'unknown', 'action', 'adventure', 'animation',\
                                    'children', 'comedy', 'crime', 'documentary', 'drama', 'fantasy',\
                                    'film_noir', 'horror', 'musical', 'mystery', 'romance', 'sci_fi',\
                                    'thriller', 'war', 'western'])
df_movies.head()

movie_id        movie_title release_date  video_release_date  \
0         1   Toy Story (1995)  01-Jan-1995                 NaN   
1         2   GoldenEye (1995)  01-Jan-1995                 NaN   
2         3  Four Rooms (1995)  01-Jan-1995                 NaN   
3         4  Get Shorty (1995)  01-Jan-1995                 NaN   
4         5     Copycat (1995)  01-Jan-1995                 NaN   

                                            imdb_url  unknown  action  \
0  http://us.imdb.com/M/title-exact?Toy%20Story%2...        0       0   
1  http://us.imdb.com/M/title-exact?GoldenEye%20(...        0       1   
2  http://us.imdb.com/M/title-exact?Four%20Rooms%...        0       0   
3  http://us.imdb.com/M/title-exact?Get%20Shorty%...        0       1   
4  http://us.imdb.com/M/title-exact?Copycat%20(1995)        0       0   

   adventure  animation  children  ...  fantasy  film_noir  horror  musical  \
0          0          1         1  ...        0          0       0        0   
1          1          0         0  ...        0          0       0        0   
2          0          0         0  ...        0          0       0        0   
3          0          0         0  ...        0          0       0        0   
4          0          0         0  ...        0          0       0        0   

   mystery  romance  sci_fi  thriller  war  western  
0        0        0       0         0    0        0  
1        0        0       0         1    0        0  
2        0        0       0         1    0        0  
3        0        0       0         0    0        0  
4        0        0       0         1    0        0  

[5 rows x 24 columns]

### Problem A
Compute the average rating for every genre of movie (for your convenience, we define the genres as a list below corresponding to the column names in the movies dataframe). Which genre is rated the highest on average? 

In [12]:
genres = ['unknown', 'action', 'adventure', 'animation', 'children', 
          'comedy', 'crime', 'documentary', 'drama', 'fantasy', 
          'film_noir', 'horror', 'musical', 'mystery', 'romance', 
          'sci_fi','thriller', 'war', 'western']

In [13]:
moviesByGenre = {}
for genre in genres:
    moviesByGenre[genre] = []

for i in range(0, len(df_movies)):
    for genre in genres:
        if df_movies[genre][i] == 1:
            moviesByGenre[genre].append(df_movies['movie_id'][i])

ratingsByGenre = {}
for genre in genres:
    ratings = []
    for movieId in moviesByGenre[genre]:
        index = list(df_ratings['movie_id']).index(movieId)
        rating = df_ratings['rating'][index]
        ratings.append(rating)
    if len(ratings) == 0: 
        ratingsByGenre[genre] = 0
    else:
        ratingsByGenre[genre] = np.mean(ratings)

topGenre = ''
topRating = -1
for key, value in ratingsByGenre.items():
    if value > topRating:
        topRating = value
        topGenre = key

print('Top Genre: ' + topGenre + '\nRating: ' + str(topRating))

Top Genre: film_noir
Rating: 3.8333333333333335


The genre with the highest average rating is 'film_noir', which has an average rating of about 3.83. To determine this information, the moviesByGenre dictionary was filled in to associate each genre with each movie that is part of that genre. Next, the rating of each movie in a given genre was found, and the average of these ratings for that genre was placed into the ratingsByGenre dictionary. Finally, the highest average rating was found from the values in this dictionary, and the associated key (genre) was the one with the highest average.

### Problem B
What is the probability that a random user likes (rates 4 or 5) at least one horror movies given that the the user likes (rates 4 or 5) at least one musical.

In [14]:
# Write your code to solve Problem B here
def doesUserLikeGenre(user, genre):
    my_user_ratings = df_ratings.loc[df_ratings['user_id'] == user]
    for key in my_user_ratings['user_id'].keys():
        if my_user_ratings['movie_id'][key] in moviesByGenre[genre]:
            if my_user_ratings['rating'][key] >= 4:
                 return True
    return False

def doesUserLikeBothGenres(user, genre_a, genre_b):
    likesA = False
    likesB = False
    my_user_ratings = df_ratings.loc[df_ratings['user_id'] == user]
    for key in my_user_ratings['user_id'].keys():
        if my_user_ratings['movie_id'][key] in moviesByGenre[genre_a]:
            if my_user_ratings['rating'][key] >= 4:
                 likesA = True
        if my_user_ratings['movie_id'][key] in moviesByGenre[genre_b]:
            if my_user_ratings['rating'][key] >= 4:
                likesB = True
        if likesA and likesB:
            return True
    return False


users = []
for i in range (0, len(df_users)):
    user = df_users['user_id'][i]
    if doesUserLikeGenre(user, 'musical'):
        users.append(user)

likesBothGenres = 0
total = 0

for i in range(0, 10001):
    total +=1
    random.shuffle(users)
    # check if user also likes comedy
    if doesUserLikeGenre(users[0], 'horror'):
        likesBothGenres +=1

print(100*(likesBothGenres/total))

76.45235476452355


The probability that a random user likes (rates 4 or 5) at least one horror movie given that the the user likes (rates 4 or 5) at least one musical is about 76%. This is information was determined by simulating many random events. In each event, a random user was selected from a list of users who liked at least one musical. From there, it was determined whether that person also liked at least one horror movie. Random users who also liked a horror movie given that they liked a musical were summed, and this sum was divided by the total number of random events to obtain the above probability.

### Problem C
What is the expected age of a random user that likes (rates a 4 or 5) at least one children's movie?

In [15]:
# Write your code to solve Problem C here
def getUsersWhoLikeGenre(genre):
    users = set()
    for i in range(0, len(df_ratings)):
        if df_ratings['movie_id'][i] in moviesByGenre[genre] and df_ratings['rating'][i] >= 4:
            users.add(df_ratings['user_id'][i])
    return users

usersWhoLikeChildMovie = list(getUsersWhoLikeGenre('children'))
expectedAges = []

for i in range(0, 10001):
    random.shuffle(usersWhoLikeChildMovie)
    user_rows = list(df_users['user_id'])
    random_user_row = user_rows.index(usersWhoLikeChildMovie[0])
    age = df_users['age'][random_user_row]
    expectedAges.append(age)

print(np.mean(expectedAges))

33.93940605939406


The expected age of a random user who likes at least one children's movie is about 34 years old. This information was determined by first creating a set of all users who liked at least one children's movie. Then, a random simulation with many events selected a random user from this list each time and recorded the user's age. Finally, the average of these recorded ages was determined to obtain the above expected age.

### Problem D
Is the probability that a random user likes (rates a 4 or 5) at least one drama independent of the probability that a random user likes (rates a 4 or 5) at least one comedy? Explain your answer.

In [16]:
# Write your code to solve Problem D here
doesUserLikeGenre(3, 'drama')

total = 0
numLikesDrama = 0
numLikesComedy = 0
numLikesBoth = 0

users = []
for i in range (0, len(df_users)):
    users.append(df_users['user_id'][i])

for i in range(0, 100001):
    total += 1
    random.shuffle(users)
    user = users[0]
    likesDrama = doesUserLikeGenre(user, 'drama')
    likesComedy = doesUserLikeGenre(user, 'comedy')
    likesBoth = doesUserLikeBothGenres(user, 'drama', 'comedy')
    if likesDrama:
        numLikesDrama += 1
    if likesComedy:
        numLikesComedy += 1
    if likesBoth: 
        numLikesBoth += 1
        
print('Probability likes drama: ' + str(100*(numLikesDrama/total)))
print('Probability likes comedy: ' + str(100*(numLikesComedy/total)))
print('Probability likes both: ' + str(100*(numLikesBoth/total)))

print('\nA and B: ' + str(100*(numLikesBoth/total)))
print('A*B: ' + str(100*((numLikesDrama/total)*(numLikesComedy/total))))

Probability likes drama: 99.6810031899681
Probability likes comedy: 97.08602913970861
Probability likes both: 96.8630313696863

A and B: 96.8630313696863
A*B: 96.77632780376629


*Write your answer to Problem D here.* 

We check whether P(A ∩ B) equals P(A) x P(B), where A is the event that the user likes at least one drama, and B is the event that the user likes at least one comedy. Because these probabilities are equal as demonstrated above with code, we can say that the two events are independent.

## Submitting HW 4
1. Double check that you have written all of your answers along with your supporting work in this notebook. Make sure you save the complete notebook.
1. Double check that your entire notebook runs correctly and generates the expected output. To do so, you can simply select Kernel -> Restart and Run All. 
2. You will download two versions of your notebook to submit, a .pdf and a .py. To create a PDF, we reccomend that you select File --> Download as --> HTML (.html). Open the downloaded .html file; it should open in your web broser. Double check that it looks like your notebook, then print a .pdf using your web browser (you should be able to select to print to a pdf on most major web browsers and operating systems). Check your .pdf for readability: If some long cells are being cut off, go back to your notebook and split them into multiple smaller cells. To get the .py file from your notebook, simply select File -> Download as -> Python (.py) (note, we recognize that you may not have written any Python code for this assignment, but will continue the usual workflow for consistency). 
3. Upload the .pdf to gradescope under hw4 report and the .py to gradescope under hw4 code. If you work with a partner, only submit one document for both of you, but be sure to add your partner using the [group feature on gradescope](https://www.gradescope.com/help#help-center-item-student-group-members).